beefy-gadget/src/gossip.rs

```diff
@@ -143,11 +146,100 @@ where
 				Err(_) => return true,
 			};
 
-			let allowed = BeefyGossipValidator::<B>::is_live(&live_rounds, msg.commitment.block_number);
+			let allowed = GossipValidator::<B>::is_live(&live_rounds, msg.commitment.block_number);
 
 			trace!(target: "beefy", "🥩 Message for round #{} allowed: {}", msg.commitment.block_number, allowed);
 
 			allowed
 		})
 	}
 }
+
+#[cfg(test)]
+mod tests {
+	use super::{GossipValidator, MAX_LIVE_GOSSIP_ROUNDS};
+	use sc_network_test::Block;
+
+	#[test]
+	fn note_round_works() {
+		let gv = GossipValidator::<Block>::new();
+
+		gv.note_round(1u64);
+
+		let live = gv.live_rounds.read();
+		assert!(GossipValidator::<Block>::is_live(&live, 1u64));
+
+		drop(live);
+
+		gv.note_round(3u64);
+		gv.note_round(7u64);
+		gv.note_round(10u64);
+
+		let live = gv.live_rounds.read();
+
+		assert_eq!(live.len(), MAX_LIVE_GOSSIP_ROUNDS);
+
+		assert!(!GossipValidator::<Block>::is_live(&live, 1u64));
+		assert!(GossipValidator::<Block>::is_live(&live, 3u64));
+		assert!(GossipValidator::<Block>::is_live(&live, 7u64));
+		assert!(GossipValidator::<Block>::is_live(&live, 10u64));
+	}
+
+	#[test]
+	fn keeps_most_recent_max_rounds() {
+		let gv = GossipValidator::<Block>::new();
+
+		gv.note_round(3u64);
+		gv.note_round(7u64);
+		gv.note_round(10u64);
+		gv.note_round(1u64);
+
+		let live = gv.live_rounds.read();
+
+		assert_eq!(live.len(), MAX_LIVE_GOSSIP_ROUNDS);
+
+		assert!(GossipValidator::<Block>::is_live(&live, 3u64));
+		assert!(!GossipValidator::<Block>::is_live(&live, 1u64));
+
+		drop(live);
+
+		gv.note_round(23u64);
+		gv.note_round(15u64);
+		gv.note_round(20u64);
+		gv.note_round(2u64);
+
+		let live = gv.live_rounds.read();
+
+		assert_eq!(live.len(), MAX_LIVE_GOSSIP_ROUNDS);
+
+		assert!(GossipValidator::<Block>::is_live(&live, 15u64));
+		assert!(GossipValidator::<Block>::is_live(&live, 20u64));
+		assert!(GossipValidator::<Block>::is_live(&live, 23u64));
+	}
+
+	#[test]
+	fn note_same_round_twice() {
```

Comment:
This is another edge case. How do we want to deal with it? We can adapt the check in `note_round()` for this case as well or just ignore it.

It only happens if we are at `MAX_LIVE_GOSSIP_ROUNDS`  and for the  **first** time a round is noted for the second time which is **not** the currently lowest numbered round.

Easiest way to fix this is to just introduce the `off-by-one` bug again and accept that in some cased, we actually keep `MAX_LIVE_GOSSIP_ROUNDS + 1 ` rounds live.